## 에러 처리

## 에러 처리의 필요성

if문, 단축 평가, 옵셔널 체이닝 연산자를 통해 확인하거나

try .. catch (…finally) 문을 사용해 발생한 에러에 적절하게 대응하면 프로그램이 강제 종료되지 않는다.

## Error 객체

Error 생성자 함수는 에러 객체를 생성한다.

에러 객체는 message 프로퍼티(인수로 전달한 문자열)와 stack 프로퍼티(에러를 발생시킨 콜 스택의 호출 정보를 나타내는 문자열)를 갖는다.

| 생성자 함수    | 인스턴스                                           |
| -------------- | -------------------------------------------------- |
| Error          | 일반적 에러 객체                                   |
| SyntaxError    | 자바스크립트 문법에 맞지 않음                      |
| ReferenceError | 참조할 수 없는 식별자 참조                         |
| TypeError      | 피연산자 또는 인수의 데이터 타입이 유효하지 않음   |
| RangeError     | 숫자값의 허용 범위를 벗어남                        |
| URIError       | encodeURI 또는 decodeURI 함수에 부적절한 인수 전달 |
| EvalError      | eval 함수에서 발생하는 에러 객체                   |

## throw 문

에러 객체 생성 ≠ 에러 발생

> 에러를 발생시키려면 try 코드 블록에서 throw 문으로 에러 객체를 던져야 한다.

`throw 표현식`

## 에러의 전파

> 에러는 호출자 방향으로 전파된다.
> 즉, 콜 스택의 아래 방향으로 전파된다.

throw된 에러를 캐치하지 않으면 호출자 방향으로 전파된다.

이때 throw된 에러를 캐치하여 적절히 대응하면 코드의 실행 흐름을 복구할 수 있다.

주의할 것은 비동기 함수 setTimeout, promise 후속 처리 메서드의 콜백 함수는 호출자가 없다는 것이다.

setTimeout, promise의 후속 처리 메서드의 콜백 함수는 태스크 큐나 마이크로태스크 큐에 일시 저장되었다가 콜 스택이 비면 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다. 이때 에러를 전파할 호출자가 없다.
